corehq/apps/users/audit/change_messages.py

```diff
@@ -0,0 +1,154 @@
+        }
+
+    @staticmethod
+    def phone_number_removed(phone_number):
+        # ToDo: dedup with phone_numbers_removed
+        return {
+            "phone_number": {"removed": [phone_number]}
+        }
+
+    @staticmethod
+    def phone_numbers_removed(phone_numbers):
+        return {
+            "phone_number": {"removed": phone_numbers}
+        }
+
+    @staticmethod
+    def profile_info(profile_id, profile_name=None):
+        if profile_id:
+            change_message = {"profile": {"id": profile_id, "name": profile_name}}
+        else:
+            change_message = {"profile": {"id": None}}
+        return change_message
+
+    @staticmethod
+    def primary_location_removed():
+        return {"location": {"id": None}}
+
+    @staticmethod
+    def commcare_user_primary_location_info(location):
+        # ToDo: Dedup with web_user_primary_location_info
+        if location:
+            change_message = {"location": {"id": location.location_id, "name": location.name}}
+        else:
+            change_message = {"location": {"id": None}}
+        return change_message
+
+    @staticmethod
+    def web_user_primary_location_info(location):
+        if location:
+            change_message = {"location": {"id": location.location_id, "name": location.name}}
+        else:
+            change_message = {"location": {"id": None}}
+        return change_message
+
+    @staticmethod
+    def commcare_user_assigned_locations_info(locations):
+        # ToDo: Dedup with web_user_assigned_locations_info
+        if locations:
+            change_message = {
+                "assigned_locations": [
+                    {'id': location.location_id, 'name': location.name}
+                    for location in locations
+                ]
+            }
+        else:
+            change_message = {"assigned_locations": []}
+        return change_message
+
+    @staticmethod
+    def web_user_assigned_locations_info(locations):
+        if locations:
+            change_message = {
+                "assigned_locations": [
+                    {'id': location.location_id, 'name': location.name}
+                    for location in locations
+                ]
+            }
+        else:
+            change_message = {"assigned_locations": []}
+        return change_message
+
+    @staticmethod
+    def groups_info(groups):
+        if groups:
+            change_message = {'groups': [
+                {'id': group.get_id, 'name': group.name}
+                for group in groups
+            ]}
+        else:
+            change_message = {'groups': []}
+        return change_message
+
+    @staticmethod
+    def added_as_web_user(domain):
+        return {"domain": {"added": domain}}
+
+    @staticmethod
+    def invited_to_domain(domain):
+        return {"domain_invitation": {"added": [domain]}}
+
+    @staticmethod
+    def invitation_revoked_for_domain(domain):
+        return {"domain_invitation": {"revoked": [domain]}}
```

Comment:
there isn't much consistency in the structure of these:
```
{name: value}
{name: value, detail: detail_value}
{name: {...}}
{name: {action: value}
{name: {action: [value]}
```

I still this a standardised format would be good:
```python
action = 'set' or 'reset' or 'remove' or 'add'   # present tense
{field_name: [{action: {details}}]}
```

```

Reply:
Hey Simon,

Thanks for the feedback here. I am still working on this PR and was definitely considering to give this a better structure from your suggestion [here](https://docs.google.com/document/d/15WS2OuB8l8UJePuwnZASIerxIrB2Gfnf1RRt24O3n8M/edit?disco=AAAANhNzkso)

Just sharing from what I had thought so far,
I see `field_name` would domain/domain_invitation/role and other slugs which sounds good.
(Also, i thought of extracting out slugs out as constants `UserChangeMessageV1.DOMAIN_SLUG = 'domain'` but that seemed like an overkill)
Standardizing action definitely sounds good. "add", "remove", "set" and "reset" sounds good.
I am still thinking through this where details would be different for different slugs. 

It could simply a boolean like for password reset
`{field_name: {'reset': True}}`
or a string
`{"domain_invitation": {"remove": domain}}`
or id/name for roles/locations/groups
`{"primary_location": {"set": {'id': location_id, 'name': location_name}}}`
 (unless we decide [not to keep names](https://docs.google.com/document/d/15WS2OuB8l8UJePuwnZASIerxIrB2Gfnf1RRt24O3n8M/edit?disco=AAAANhNzksk) and fetch them when making message, then this would be just "id")

We could also do 
`{"domain_invitation": {"remove": {'domain': domain}}}` to make second more like 3 but that looks more complex than needed.

Its not clear why this would be a list though, as in your suggestion. If there are multiple actions it would be something like
`{"domain_invitation": {"remove": [domain], "add": [domain]}}`

Let me know what you think.

Thanks again for proactive review on this.

Reply:
The last suggestion seems fine. I guess it doesn't matter whether the list is of action dicts or action details:
```json
{"domain_invitation": {"remove": [domain], "add": [domain]}}
vs
{"domain_invitation": [{"remove": domain}, {"add": domain}]}
```

Reply:
Hey Simon,

So I went through few possibilities and eventually came up with the following structure

https://github.com/dimagi/commcare-hq/pull/30181/commits/34f0fb4815e6f0acee2d59d6151e35a946ca8871#diff-4284bb62f43ca4b0545a51692cec107399e2d26d955097a6893752c50efa4609R13-R25

also included dict as an option for changes that *should* have only one change like primary location and then list for where there can be one or more like assigned locations.
Updated where detail could be a string like `{"domain_invitation": {"remove": domain}}` to `{"domain_invitation": {"remove": {"domain": domain}}` to match dict structure instead because it should have only one entry (domain in this example).

Let me know your thoughts on this.

Reply:
the problem with having different data structures is you need to know what the field and action are before you can interpret the details, there is no way to process them consistently without checking types. Having said that even if we made the data structure consistent there is no consistency in the details fields so maybe it doesn't matter.

Reply:
I was trying to have a common structure, but then it was forcing simple things to be made complicated.
Some logs are very simple like `{'password': 'reset'}` which was made `{"password": {'reset': True}}` to bring some structure but something like assigned locations will always have a more complex structure. So there is some variety there that can't be put together.
All related model changes(role/location/groups/program/profile) follow the same pattern of id + name but other fields need flexibility to store what they have/need.

Yes, you are right,
The class that interprets these message, would get set/reset/add/remove as the action values only, but it would need to know the structure for detail fields for the field + action, though it should not really change over time once set and agreed upon.

Plus, This class and the one that parses the messages should be considered frozen (hence versioned) unless there are simple changes related to just display that don't need to add more info to logs like using "project" instead of "domain" in messages or adding links to roles/locations instead of `Name[ID]` structure which should be okay.